src/v/cluster/scheduling/leader_balancer_constraints.h

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "cluster/partition_leaders_table.h"
+#include "cluster/scheduling/leader_balancer_types.h"
+#include "cluster/topic_table.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "raft/types.h"
+
+#include <seastar/core/metrics.hh>
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/btree_map.h>
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+#include <cstdint>
+#include <functional>
+#include <limits>
+#include <numeric>
+#include <optional>
+#include <unordered_map>
+
+namespace cluster::leader_balancer_types {
+
+class shard_index final : public index {
+public:
+    explicit shard_index(index_type index)
+      : _index(std::move(index)) {}
+
+    shard_index(shard_index&& si) noexcept
+      : _index(std::move(si._index)) {}
+    shard_index(const shard_index& si) noexcept
+      : _index(si._index) {}
```

Comment:
why don't the default versions work here? also, can we delete the copy versions?

Reply:
The copy constructors are used for unit tests. And yeah great point, switching to default implementations of the constructors.